src/glue/jobs/s3_to_json.py

```diff
@@ -16,6 +16,237 @@
+        transformed_array = transform_object_to_array_of_objects(
+            json_obj,
+            key_name='OffsetInSeconds',
+            key_type=int,
+            value_name='HeartRate',
+            value_type=int
+        )
+
+        # Resulting `transformed_array`:
+        # [
+        #     {'OffsetInSeconds': 0, 'HeartRate': 62},
+        #     {'OffsetInSeconds': 1, 'HeartRate': 63},
+        #     {'OffsetInSeconds': 2, 'HeartRate': 62},
+        # ]
+    """
+    array_of_obj = [
+            {
+                key_name: key_type(k),
+                value_name: value_type(v)
+            }
+            for k, v in json_obj_to_replace.items()
+    ]
+    return array_of_obj
+
+def transform_json(
+        json_obj: dict,
+        dataset_identifier: str,
+        metadata: dict,) -> dict:
+    """
+    Perform the following transformations:
+
+    For every JSON:
+        - Add an export_start_date property (may be None)
+        - Add an export_end_date property (may be None)
+
+    For JSON whose data types have a subtype:
+        - Add subtype as "Type" property
+
+    For JSON whose data type is "EnrolledParticipants":
+        - Cast "CustomFields.Symptoms" and "CustomFields.Treatments" property
+          to an array. (May be an empty array).
+
+    For relevant Garmin data types:
+        - Some Garmin data types have one or more properties which are objects
+          (usually mapping time to some metric), that would be better formatted
+          as an array of objects. We transform these properties into arrays of
+          objects.
+
+    Args:
+        json_obj (str): A JSON object sourced from the JSON file of this data type.
+        dataset_identifier (str): The data type of `json_obj`.
+        metadata (dict): Metadata derived from the file basename.
+
+    Returns:
+        json_obj (dict) The JSON object with the relevant transformations applied.
+    """
+    logger = logging.getLogger(__name__)
+    if metadata["start_date"] is not None:
+        json_obj["export_start_date"] = metadata["start_date"].isoformat()
+    else:
+        json_obj["export_start_date"] = None
+    json_obj["export_end_date"] = metadata.get("end_date").isoformat()
+    if dataset_identifier in DATA_TYPES_WITH_SUBTYPE:
+        # This puts the `Type` property back where Apple intended it to be
+        json_obj["Type"] = metadata["subtype"]
+    if dataset_identifier == "SymptomLog":
+        # Load JSON string as dict
+        json_obj["Value"] = json.loads(json_obj["Value"])
+    if dataset_identifier == "EnrolledParticipants":
+        for field_name in ["Symptoms", "Treatments"]:
+            if (
+                    field_name in json_obj["CustomFields"]
+                    and isinstance(json_obj["CustomFields"][field_name], str)
+               ):
+                if len(json_obj["CustomFields"][field_name]) > 0:
+                    # This JSON string was written in a couple different ways
```

Comment:
Wow, this is nasty. Is this worth letting care evolution know about this?

Reply:
I haven't seen the `[{\\\"id\\\": ...` format in the production data (we would log a warning if we ran into a `JSONDecodeError`), so whatever issues they had with the pilot data seem to have been resolved.

Reply:
We just have to make sure we are monitoring the logs for these warnings.